src/full/Agda/TypeChecking/CheckInternal.hs

```diff
@@ -160,13 +180,17 @@ instance CheckInternal Term where
         s <- shouldBeSort t
         reportSDoc "tc.check.internal" 30 $ "pi type should have sort" <+> prettyTCM s
         when (s == SizeUniv) $ typeError $ FunctionTypeInSizeUniv v
+        experimental <- optExperimentalIrrelevance <$> pragmaOptions
         let sa  = getSort a
             sb  = getSort (unAbs b)
             mkDom v = El sa v <$ a
             mkRng v = fmap (v <$) b
             -- Preserve NoAbs
-            goInside = case b of Abs{}   -> addContext (absName b, a)
-                                 NoAbs{} -> id
+            goInside = case b of
+              Abs{}   -> addContext $ (,)
+                (absName b)
+                ((if experimental then mapRelevance irrToNonStrict else id) a)
```

Comment:
`applyWhen experimental (mapRelevance irrToNonStrict) a`

Reply:
The extent of my haskell skills has been revealed. Thank you for suggesting this more idiomatic style, I wasn't happy with it either.